chore(#1297): Update CHT Sync doc Part 1 (#1300)
medic/cht-docs: content/en/apps/guides/data/analytics/setup.md

---
title: "CHT Sync Setup"
weight: 100
linkTitle: "Local Setup"
description: >
  Setting up CHT Sync with the CHT
relatedContent: >
  core/overview/architecture
  core/overview/cht-sync
---

> {{% pageinfo %}}
> These instructions apply to both CHT 3.x (beyond 3.12) and CHT 4.x.  
> {{% /pageinfo %}}
> 
> Medic maintains CHT Sync which is an integrated solution designed to enable data synchronization between CouchDB and PostgreSQL for the purpose of analytics. It can easily be deployed using Docker. It is supported on CHT 3.12 and later, including CHT 4.x. By using CHT Sync a CHT deployment can easily get analytics by using a data visualization tool. All tools are open source and have no licensing fees.
> 
> CHT Sync has been designed to work in both local development environments for testing models or workflows, and in production environments. The setup can accommodate the needs of different stages or environments.
> 
> ### Prerequisites
> 
> - [Docker](https://docs.docker.com/install/)
> - [Docker Compose](https://docs.docker.com/compose/install/)
> - [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
> - An `.env` file containing the environment variable. Placeholders can be found in the `env.template` file. The `.env` file should be located in the root directory of the project or set by the operating system. The variables should be customized accordingly for the specific deployment needs.
> 
> {{% alert title="Note" %}}
> In order for CHT Sync to transform CouchDB data from the CHT in an easier to query format, it needs to be linked to [CHT Pipeline](https://github.com/medic/cht-pipeline), which contains transformation models for DBT. The schema differs from `couch2pg`. See [`./postgres/init-dbt-resources.sh`](https://github.com/medic/cht-sync/blob/main/postgres/init-dbt-resources.sh).
> {{% /alert %}}

Reviewer: I think this should be moved to the intro page and we should re-title the that page to "Introduction and Prerequisites".  The prereqs for dev/local an production will be the same.

Instead, here we should focus on why you would want a local development setup and what the differences are from production.  I'll likely get this wrong, but here's my suggestion:

```suggestion
Before setting up CHT Sync in production, it's very handy to be able to run it locally.  This will allow you to experiment with the data flow and easily query development data quickly and locally. 

These instructions assume you're running CHT Sync, CHT Core and Postgres either locally on your workstation or on a local server.  They are not meant to be used to deploy a secure, always on production instance.
```

> #### Environment variables

Reviewer: This whole section should be it's own page.  It is 100% shared between production and development, so should not be only on the local development page or only on the prod page.  

also, by removing this section it will greatly simplify the install instructions and reduce the "wall of text" fear that first time readers will have.  Further, we can suggest which _just_ the `.env` vars to change applicable for development.  in the install instructions.

There are three environment variable groups in the `.env` file. To successfully set up CHT Sync, it is important to understand the difference between them.
1. `POSTGRES_`: Used by PostgREST and PostgreSQL to establish the PostgreSQL database to synchronize CouchDB data to. They also define the schema and table names to store the CouchDB data. The main objective is to define the environment where the raw CouchDB data will be copied.
2. `DBT_`: Exclusive to the DBT configuration. The main objective is to define the environment where the tables and views for the models defined in `CHT_PIPELINE_BRANCH_URL` will be created. It is important to separate this environment from the previous group. `DBT_POSTGRES_SCHEMA` must be different from `POSTGRES_SCHEMA`. `DBT_POSTGRES_HOST` has to be the Postgres instance created with the environment variables set in the first group.
3. `COUCHDB_`: Used by CouchDB and Logstash to define the CouchDB instance to sync with. With `COUCHDB_DBS`, we can specify a list of databases to sync.

> All the variables in the `.env` file:

Reviewer: love LOVE the detail here !  So helpful have not only then name, but also the default value and the description.  nice work!

| `COUCHDB_PASSWORD`        | `password`                                         | Password of the CouchDB instance to sync with                                                                                                  |
| `COUCHDB_DBS`             | `"medic"`                                          | Space separated list of databases to sync e.g `"medic medic_sentinel"`                                                                         |
| `COUCHDB_HOST`            | `couchdb`                                          | Host of the CouchDB instance to sync with                                                                                                      |
| `COUCHDB_PORT`            | `5984`                                             | Port of the CouchDB instance to sync with                                                                                                      |
| `COUCHDB_SECURE`          | `false`                                            | Is connection to CouchDB instance secure?                                                                                                      |

{{% alert title="Note" %}}

> If `CHT_PIPELINE_BRANCH_URL` is pointing to a private repo then you need to provide an access token in the url i.e. `https://<PAT>@github.com/medic/cht-pipeline.git#main`. In this example you will replace `<PAT>`  with an access token from Github. Instruction on how to generate one can be found [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens).
> {{% /alert %}}

Reviewer: Wording and simplification.  Don't use `here` for a link, but the actual description of where you're going.  `Github` is `GitHub`:

```suggestion
If `CHT_PIPELINE_BRANCH_URL` is pointing to a private GitHub repository, you'll need an access token in the URL. Assuming your repository is `medic/cht-pipeline`, you would replace  `<PAT>`  with an access token: `https://<PAT>@github.com/medic/cht-pipeline.git#main`. Please see [GitHub's instructions](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) on how to generate a token.
{{% /alert %}}
```

Author: This is so much better, thank you for the suggestion!

> 

Reviewer: We describe 3 setups here - awesome!  However, I wasn't actually told what to do when I was done. Please add some more content for this. Suggestions might be how to verify the command worked:

> You can verify this command worked by running `docker ps`.  It should show 6 containers running including Logstash, DBT, data generator, Postgres, CouchDB and Postgrest (note the `t` at the end!).  

As well, it's not clear you know how long to wait.  Maybe something like:

> The first time you run this it will need to download many docker images and will take a while.  You'll know it's done when you see `#8 DONE 0.0s` and your returned to the command line.  Be patient!

It wasn't actually clear to me what I'd achieved after running a command.  Suggest adding some next steps.  This might be:

> Now that all services are running, use a Postgres client like [Beekeeper](https://www.beekeeperstudio.io/) to connect to server `localhost:5432` with user `postgres` and password `postgres`.  You should be able to see sample data being inserted into the `medic_is_awesome` table.

Finally, we should mention how to clean things up for each scenario too. Maybe something like:

> When you're done using the services, you can clean everything by running`down`. Add `-v` at the to  remove all your data volumes: 
> 
> ```
> docker compose -f docker-compose.couchdb.yml -f docker-compose.postgres.yml -f docker-compose.yml down
> ```

> ### Setup

Reviewer: ```suggestion
## Setup
```

> The following configurations facilitate data synchronization, transformation, and storage for local development and testing.

Reviewer: with simplifications to this page made above, this intro text is no longer needed.

```suggestion
```

> Copy the values in `env.template` file to the `.env` file and update them accordingly to the local configuration for the different scenarios below.

Reviewer: This is not sufficient instructions for setup.  Please enumerate the specific `.env` vars we need to change and specify example values.

EG:

> Assuming you're running on your laptop with IP `192.168.0.2` and a both CHT Core and Postgres are running locally, change these three variables:
>
> * `POSTGRES_HOST` - set to `192.168.0.2`
> * `DBT_POSTGRES_HOST` - set to `postgres-local`
> * `COUCHDB_HOST` - set to `couchdb-local`
>
> Leave all other `.env` var values unchanged.

> Install the dependencies:

Reviewer: Nice work on this!  I first skipped this step and got stuck with an error `sh: 1: ts-node: not found` .  Then read more closely and this fixed it \o/

```sh
npm install
```

#### With sample CouchDB data
This setup involves starting Logstash, PostgreSQL, PostgREST, DBT, and CouchDB. Sample fake data is generated for CouchDB.

Run the Docker containers and wait for every container to be up and running:
```sh

> # starts logstash, postgres, postgrest, generator (for fake data), couchdb and dbt
> npm run local

Reviewer: We just told them what this command will do, we can simplify by removing duplicate content:

```suggestion
npm run local
```

Also - is there a reason this command gets a special `npm` entry while the rest are all raw `docker compose` calls?  Maybe remove the use of `npm` here so there's parity among all the calls to all be `docker compose`?

```

> This is equivalent to running:
> ```sh
> # starts logstash, postgres, postgrest, generator (for fake data), couchdb and dbt
> docker-compose -f docker-compose.couchdb.yml -f docker-compose.postgres.yml -f docker-compose.yml up -d
> ```

Reviewer: I think this is too much information and might be confusing to someone who wonders if they should run this, as we're presenting it in the "copy an paste this code" format.

Instead, recommend a deep link:

```suggestion
This is a wrapper for some `docker compose` calls.  To find out exactly what is being called, see [the source script](https://github.com/medic/cht-sync/blob/main/scripts/config.ts#L40). 
```

Author: I actually removed the `npm run local` instruction and kept only the docker compose calls, for the sake of consistency. @witash is working on removing the node script.

> #### With a separate CouchDB instance 

Reviewer: ```suggestion
### Separate CouchDB instance 
```

This setup involves starting Logstash, PostgreSQL, PostgREST, and DBT. It assumes you have a CouchDB instance running, and you updated the `.env` CouchDB variables accordingly.

Run the Docker containers locally and wait for every container to be up and running:
```sh

> # starts logstash, postgres, postgrest, and dbt
> docker-compose -f docker-compose.postgres.yml -f docker-compose.yml up -d

Reviewer: We just told them what this command will do, we can simplify by removing duplicate content:

```suggestion
docker-compose -f docker-compose.postgres.yml -f docker-compose.yml up -d
```

```

> #### With separate CouchDB and PostgreSQL instances

Reviewer: ```suggestion
### Separate CouchDB and PostgreSQL instances
```

This local setup involves starting Logstash, PostgREST, and DBT. It assumes that CouchDB and PostgreSQL instances are run separately from the Docker Compose provided with CHT Sync, and the `.env` variables were updated to match those instances details.

Run the Docker containers locally and wait for every container to be up and running:
```sh

> # starts logstash, postgrest, and dbt
> docker-compose -f docker-compose.postgrest.yml -f docker-compose.yml up -d logstash postgrest dbt

Reviewer: We just told them what this command will do, we can simplify by removing duplicate content:

```suggestion
docker-compose -f docker-compose.postgrest.yml -f docker-compose.yml up -d logstash postgrest dbt
```

```

> ### Setup Superset

Reviewer: ```suggestion
## Setup Superset
```
